File: programmers/kakao#2/python/KAKAO29.py
from itertools import permutations

def solution(n, weak, dist):
    INF = float('inf')
    min_count = INF
    
    for start in range(len(weak)):
        for d in permutations(dist):
            current = start
            cnt = 1
            for i in range(len(weak)):
                _next = (start + i) % len(weak)
                diff = weak[_next] - weak[current]
                
                if _next < current:
                    diff += n
                
                if diff > d[cnt - 1]:
                    current = _next
                    cnt += 1
                    if cnt > len(dist) or cnt >= min_count:
                        break
                
            if cnt <= len(dist) and cnt < min_count:
                min_count = cnt
                
    return min_count if min_count != INF else -1

File: programmers/kakao#2/java/KAKAO29.java
import java.util.*;
class KAKAO29 {
    static final int INF = Integer.MAX_VALUE;
    int result = INF;
    public int solution(int n, int[] weak, int[] dist) {
        Arrays.sort(dist);
        
        List<Integer> list = new ArrayList<>();
        
        for (int w : weak) {
            list.add(w);
        }
        permutate(n, list, dist, 0);
        return result == INF ? -1 : result;
    }
    
    public void permutate(int n, List<Integer> weaks, int[] dist, int count) {
        if (weaks.isEmpty() && count < result) {
            result = count;
            return;
        }
        
        if (count >= dist.length)
            return;

        for (int i=0;i<weaks.size();i++) {
            List<Integer> remains = new ArrayList<>();
            int start = weaks.get(i);
            int end = start + dist[dist.length - count - 1];
            
            for (int j=0;j<weaks.size();j++) {
                if (weaks.get(j) >= start && weaks.get(j) <= end)
                    continue;
                if (weaks.get(j) < start && weaks.get(j) + n <= end)
                    continue;
                remains.add(weaks.get(j));
            }
            permutate(n, remains, dist, count + 1);
        }
    }

    
}
/**
 * [카카오 2020 공채] 외벽 점검
 * 해결: O
 * 시간: 50분
 */